includes/class-newspack-ads-gam.php

```diff
@@ -0,0 +1,386 @@
+			$network = self::get_gam_network();
+			$ad_unit->setParentId( $network->getEffectiveRootAdUnitId() );
+			$ad_unit->setTargetWindow( AdUnitTargetWindow::BLANK );
+		}
+
+		$ad_unit->setName( $name );
+
+		$ad_unit_sizes = [];
+		foreach ( $sizes as $size_spec ) {
+			$size = new Size();
+			$size->setWidth( $size_spec[0] );
+			$size->setHeight( $size_spec[1] );
+			$size->setIsAspectRatio( false );
+			$ad_unit_size = new AdUnitSize();
+			$ad_unit_size->setSize( $size );
+			$ad_unit_size->setEnvironmentType( EnvironmentType::BROWSER );
+			$ad_unit_sizes[] = $ad_unit_size;
+		}
+		$ad_unit->setAdUnitSizes( $ad_unit_sizes );
+
+		return $ad_unit;
+	}
+
+	/**
+	 * Update Ad Unit.
+	 *
+	 * @param object $ad_unit_config Ad Unit configuration.
+	 * @return AdUnit Updated AdUnit.
+	 */
+	public static function update_ad_unit( $ad_unit_config ) {
+		try {
+			$inventory_service = self::get_gam_inventory_service();
+			$found_ad_units    = self::get_gam_ad_units( [ $ad_unit_config['id'] ] );
+			if ( empty( $found_ad_units ) ) {
+				return new WP_Error( 'newspack_ads', __( 'Ad Unit was not found.', 'newspack-ads' ) );
+			}
+			$result = $inventory_service->updateAdUnits(
+				[ self::modify_ad_unit( $ad_unit_config, $found_ad_units[0] ) ]
+			);
+			if ( empty( $result ) ) {
+				return new WP_Error( 'newspack_ads', __( 'Ad Unit was not updated.', 'newspack-ads' ) );
+			}
+			return $result[0];
+		} catch ( \Exception $e ) {
+			return [];
+		}
+	}
+
+	/**
+	 * Create a GAM Ad Unit.
+	 *
+	 * @param object $ad_unit_config Configuration of the ad unit.
+	 * @return AdUnit Created AdUnit.
+	 */
+	public static function create_ad_unit( $ad_unit_config ) {
+		try {
+			$network           = self::get_gam_network();
+			$inventory_service = self::get_gam_inventory_service();
+			$ad_unit           = self::modify_ad_unit( $ad_unit_config );
+			$created_ad_units  = $inventory_service->createAdUnits( [ $ad_unit ] );
+			if ( empty( $created_ad_units ) ) {
+				return new WP_Error( 'newspack_ads', __( 'Ad Unit was not created.', 'newspack-ads' ) );
+			}
+			return self::serialize_ad_unit( $created_ad_units[0] );
+		} catch ( \Exception $e ) {
+			return [];
+		}
+	}
+
+	/**
+	 * Archive a single GAM Ad Unit.
+	 *
+	 * @param int $id Id of the ad unit to archive.
+	 */
+	public static function archive_ad_unit( $id ) {
+		try {
+			$action            = new ArchiveAdUnitsAction();
+			$inventory_service = self::get_gam_inventory_service();
+
+			$statement_builder = self::get_serialised_gam_ad_units_statement_builder( [ $id ] );
+			$result            = $inventory_service->performAdUnitAction(
+				$action,
+				$statement_builder->toStatement()
+			);
+			if ( null !== $result && $result->getNumChanges() > 0 ) {
+				return true;
+			} else {
+				return false;
+			}
+		} catch ( \Exception $e ) {
+			return false;
+		}
+	}
+
+	/**
+	 * Is GAM connected?
+	 *
+	 * @return boolean True if GAM is connected.
+	 */
+	public static function is_connected() {
```

Comment:
> I think we need to also check whether the OAuth app is accessible instead of just checking for the credentials.

Good catch - added some changes in a4c7e43fb3d12b93d97e33014f3b0d875889d7b9. Now the backend will try to get the network code to verify the connection.